rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,667 @@
+
+#### WidgetProps
+
+Another addition is the `Props` associated type:
+
+```rust
+trait Widget {
+  type Props: WidgetProps;
+  // ...
+}
+```
+
+The `WidgetProps` trait will look something like this:
+
+```rust
+trait WidgetProps: Debug {
+  fn get_on_event(&self) -> Option<OnEvent>;
+  fn get_styles(&self) -> Option<Style>;
+  fn get_focusable(&self) -> Option<bool>;
+  // ...
+}
+```
+
+##### Deriving
+
+One additional feature for `WidgetProps` this RFC would like to propose is the ability to derive `WidgetProps`. This will make it much simpler for users to define their props and for defining Kayak-specific functionality.
+
+The derive macro would have some associated macros, which will be used to mark certain fields for use in the implementation. These are:
+
+* `#[props(OnEvent)]` - Used to mark a field as the `OnEvent` prop
+* `#[props(Styles)]` - Used to mark a field as the `Styles` prop
+* `#[props(Focusable)]` - Used to mark a field as the `Focusable` prop
+
+There may be more added to this list in the future, but for now this is the main assortment.
+
+There should only be a single usage of each of these markers within a struct. If there are more than one, only the one defined last will be used (last was chosen so we don't have to check if it already exists in the macro function).
+
+If any of these attributes are missing, the relevant method will return none.
+
+Additionally, they should be allowed to be specified as `Optional`. 
+
+```rust
+#[derive(WidgetProps)]
+struct MyWidgetProps {
+  #[props(OnEvent)]
+  event_handler: OnEvent
+  #[props(Focusable)]
+  focusable: Optional<bool>,
+  // Defined without the marker attribute:
+  styles: Styles,
+}
+
+// Generated:
+impl WidgetProps {
+  fn get_on_event(&self) -> Option<OnEvent> {
+  	Some(self.event_handler.clone())  
+  }
+  
+  fn get_styles(&self) -> Option<Style> {
+    None
+  }
+  
+  fn get_focusable(&self) -> Option<bool> {
+    self.focusable
+  }
+  
+  // ...
+}
+```
+
+#### The Constructor
+
+Another addition to `Widget` would be the constructor method. This associated method will allow widgets to be generated with a set of props in a more controlled and freeform way. The method looks like this:
+
+```rust
+fn constructor(id: Index, children: Children, mut props: Self::Props) -> dyn Widget<Props=Self::Props> where Self: Sized;
+```
+
+> Note that this would put the `Sized` restriction on widgets.
+
+Doing it this way allows the user to define the naming and placement of their props, or even replace them with their own.
+
+---
+
+### Defining Widgets
+
+#### Custom Widgets
+
+A custom widget is a widget defined manually by a user using standard Rust syntax (i.e. not by using the `#[widget]` attribute). By defining a widget manually, a user has finer control over behavior, data, and more.
+
+To define a custom widget, we need to first define our struct:
+
+```rust
+#[derive(Debug)]
+struct MyButton {
+  my_id: Index,
+  my_children: Children,
```

Comment:
I think this can move into the props as well unless there is a specific reason we don't want to?

Reply:
Nope. I'll do that!